File: Abstract Data Type/pa3/List.java
//Zhijun Yang
//zyang100
//pa3

public class List{

	private class Node{
		
		Object data;
		Node next;
		Node prev;

		
		Node(Object data) {
			this.data = data;
			next = null;
			prev = null;
		}

		public String toString(){
			return String.valueOf(data);
		}

		Node(Object data, Node nextNode, Node prevNode){
			this.data = data;
			next = nextNode;
			prev = prevNode;
		}
	}

	//Fields
	private Node back;
	private Node front;
	private int length = 0;
	private Node cursor;
	private int index;

	//Constructor
	List(){
		front = back = cursor = null;
		length = 0;
		index = -1;
	}
	// Access Functions --------------------------------------------------------
	// Returns the number of elements in this List
	public int length(){
		return length;
	}

	// If cursor is defined, returns the index of the cursor element,
 	// otherwise returns -1.
 	int index(){
 		if (cursor != null){
 			return index;
 		}
 		else{
 			return -1;
 		}
 	}

 	// Returns front element. Pre: length()>0
 	Object front(){
 		if (length > 0){
 			return front.data;
 		}else{
 			throw new RuntimeException("List Error: front() called on empty List");
 		}
 	}

 	// Returns back element. Pre: length()>0
 	Object back(){
 		if(length > 0){
 			return back.data;
 		}else{
 			throw new RuntimeException("List Error: front() called on empty List");
 		}
 	}

 	// Returns cursor element. Pre: length()>0, index()>=0
 	Object get() {
 		if (length() > 0 && index() >= 0){
 			return cursor.data;
 		}
 		else{
 			return -1;
 		}

	}

 	// Returns true if and only if this List and L are the same
 	// integer sequence. The states of the cursors in the two Lists
 	// are not used in determining equality.
 	boolean equals(List L){
 		boolean eq = true;
 		Node a = this.front;
 		Node b = L.front;

 		if(this.length == L.length){
 			while(eq && a !=  null){
 				eq = (a.data == b.data);
 				a = a.next;
 				b = b.next;
 			}
 			return eq;
 		}
 		else{
 			return false;
 		}
 	}

 	// Manipulation procedures 
 	// Resets this List to its original empty state.
 	void clear(){
 		front = back = cursor = null;
 		length = 0;
 		index = -1;
 		
 	}

 	void moveFront(){
 		/*if (length <= 0){
 			throw new RuntimeException("List Error: movefront() called on empty List");
 		}
 		else{*/
 		if (length > 0){
 			cursor = front;
 			index = 0;
 		}
 	}

 	void moveBack(){
 		/*if (length <= 0){
 			throw new RuntimeException("List Error: moveBack() called on empty List");
 		}
 		else*/
 		if (length > 0){
 			cursor = back;
 			index = length - 1;
 		}
 	}

	// If cursor is defined and not at front, moves cursor one step toward
 	// front of this List, if cursor is defined and at front, cursor becomes
	// undefined, if cursor is undefined does nothing.
 	void movePrev(){
 		if (cursor != null && index == 0){
 			cursor = null;
 			index = -1;
 		}
 		else if (cursor != null && index != 0) {
 			cursor = cursor.prev;
 			index--;
 		}
 	}

 	// If cursor is defined and not at back, moves cursor one step toward
 	// back of this List, if cursor is defined and at back, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void moveNext(){
		if (cursor != null && cursor == back ){
 			cursor = null;
 			index = -1;
 		}
 		else if (cursor != null && cursor != back) {
 			cursor = cursor.next;
 			index++;
 		}
 		else if (cursor == null){
 			cursor = null;
 			index = -1;
 		}
 	}

	// Insert new element into this List. If List is non-empty,
 	// insertion takes place before front element.
 	void prepend(Object data){
 		if(length <= 0){
 			Node N = new Node(data);
			front = back = N;
 		}else{
 			Node N = new Node(data, front ,null);
 			front.prev = N;
 			front = N;
 		}
 		length++;
 		index++;
 	}

 	// Insert new element into this List. If List is non-empty,
 	// insertion takes place after back element.
 	void append(Object data){
 		if(length <= 0){
 			Node N = new Node(data);
 			front = back = N;
 		}
 		else{
 			Node N = new Node(data, null, back);
 			back.next = N; 
 			back = N;
 		}
 		length++;
 	}

 	// Insert new element before cursor.
 	// Pre: length()>0, index()>=0
 	void insertBefore(Object data){
 		Node N = new Node(data);
 		if(length() <= 0 || index() < 0){
 			throw new RuntimeException("List Error: insertBefore() called on empty List");
 		}
 		if(cursor == front){
 			prepend(data);
 		}
 		else{
 			N.prev = cursor.prev;
 			N.next = cursor;
 			cursor.prev.next = N;
 			cursor.prev = N;
 			length++;
 			index++;
 		}
 	}

 	// Inserts new element after cursor.
 	// Pre: length()>0, index()>=0
 	void insertAfter(Object data){
 		Node N = new Node(data);
 		if(length() <= 0 || index() < 0){
 			throw new RuntimeException("List Error: insertAfter() called on empty List");
 		}
 		if (index() == length()-1){
 			append(data);
 		}
 		else{
 			N.prev = cursor;
 			N.next = cursor.next;
 			cursor.next.prev = N;
 			cursor.next = N;
 			length++;
 		}
 	}

 	 // Deletes the front element. Pre: length()>0
 	void deleteFront() {
 	if (length() == 0){
 		throw new RuntimeException("List Error: deleteFront() called on empty List");
 	}else{
 		if (length() == 1){
 			clear();
 		}
 		else{
			if(cursor == front){
				cursor = null;
				index =-1;
			} else if (cursor != null){
				index--;
			}
			front = front.next;
			//front.prev.next = null;
			front.prev = null;
			length--;
		}
 	}
}

 	// Deletes the back element. Pre: length()>0
 	void deleteBack(){
 		if(length() <= 0){
 			throw new RuntimeException("List Error: deleteBack() called on empty List");
 		}
 		if(length == 1) {
			clear();
		} else if(length > 0) {
			if(cursor == back){
				index =-1;
			}
			back = back.prev;
			if(back != null) {
				back.next = null;
			}
			length--;
		} else {
			back = null;
		}
	}

 	// Deletes cursor element, making cursor undefined.
 	// Pre: length()>0, index()>=0
 	void delete() {
 		if (length() <= 0){
 			throw new RuntimeException("List Error: delete() called on empty List");
 		}
 		if (index < 0){
 			throw new RuntimeException("List Error: delete() called on empty index");
 		}
		if(cursor == front){
			deleteFront();
			return;
		} else if (cursor == back){
			deleteBack();
			return;
		} else 
		{
			cursor.next.prev = cursor.prev;
			cursor.prev.next = cursor.next;
			cursor = null;
			index = -1;
			length --;
		}
	}


 	// Other methods 
 	// Overrides Object's toString method. Returns a String
 	// representation of this List consisting of a space
 	// separated sequence of integers, with front on left.
 	public String toString(){
 		String str = "";
 		for (Node N = front; N != null; N = N.next){
 			str += N.toString() + " ";
 		}
 		return str;
 	}

 	// Returns a new List representing the same integer sequence as this
 	// List. The cursor in the new list is undefined, regardless of the
 	// state of the cursor in this List. This List is unchanged.
 	 public List copy(){
        List newList = new List();
        if (length() > 0)
        {
            Node N = front;
            while (N != null)
            {
                newList.append(N.data);
                N = N.next;
            }
        }
        newList.cursor = null;
        newList.index = -1;
        return newList;
    }

}

File: Abstract Data Type/Lab/Lab1/List.c
//Zhijun Yang
//zyang100
//pa4

#include<stdio.h>
#include<stdlib.h>
#include"List.h"

typedef struct NodeObj{
	int data;
	struct NodeObj* prev;
	struct NodeObj* next;
}NodeObj;

typedef NodeObj* Node;

typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int index;
	int length;
}ListObj;

void freeNode(Node *n){
	if (n != NULL && *n != NULL){
		free(*n);
		*n = NULL;
	}
}

Node newNode(int data){
	Node newNode = malloc(sizeof(NodeObj));
	newNode->data = data;
	newNode->next = NULL;
	newNode->prev = NULL;
	return (newNode);
}

List newList(void){
	List L;
	L = malloc(sizeof(ListObj));
	L->front = NULL;
	L->back = NULL;
	L->cursor = NULL;
	L->length = 0;
	L->index = -1;
	return(L);
}

void freeList(List* pL){
	if (pL != NULL && *pL != NULL){
		clear(*pL);
		free(*pL);
		*pL = NULL;
	}
}

int length(List L){
	if (L != NULL){
		return (L->length);
	}else{
		printf("List error: length() is NULL");
		exit(1);
	}
}

int index(List L){
	if (L != NULL){
		return (L->index);
	}else{
		printf("List error: index() is NULL");
		exit(1);
	}
}

int front(List L){
	if (L != NULL){
		return (L->front->data);
	}else{
		printf ("List error: front() is NULL");
		exit(1);
	}
}

int back(List L){
	if (L != NULL){
		return (L->back->data);
	}else{
		printf("List error: back() is NULL");
		exit(1);
	}
}

int get(List L){
	if (L == NULL){
		printf("List error: Null list");
		exit(1);
	}
	if (length(L) <= 0){
		printf("List error: get() is NULL");
		exit(1);
	}
	if (L->index == -1){
		printf("List error: index() is NULL");
		exit(1);
	}
		return L->cursor->data;
}

int equals(List A, List B){
	if (A == NULL || B == NULL){
		printf("List error: equals() is NULL");
	}
	Node aNode = A->front;
	Node bNode = B->front;

	if (length (A) != length(B))
		return 0;
	while (aNode != NULL){
		if (aNode->data != bNode->data){
			aNode = bNode = NULL;
			return 0;
		}
		else{
			aNode = aNode->next;
			bNode = bNode->next;
		}
	}
	return 1;
}

void clear(List L){
	L->front = L->back = L->cursor = NULL;
	L->length = 0;
	L->index = -1;
}

void moveFront(List L){
	if(L == NULL){
		printf("List error: moveFront() is NULL");
		exit(1);
	}
	if (L->length > 0){
		L->cursor = L -> front;
		L->index = 0;
	}
}

void moveBack(List L){
	if (L == NULL){
		printf("List error: moveBack() is NULL");
		exit(1);
	}
	if (L->length > 0){
		L->cursor = L->back;
		L->index = L->length - 1;
	}
}

void movePrev(List L){
	if(L == NULL){
		printf("List error: movePrev is NULL");
		exit(1);
	}
	if (L->cursor != NULL && L->cursor != L->front){
		L->cursor = L->cursor -> prev;
		L->index--;
	}
	else if (L->cursor != NULL && L->cursor == L->front){
		L->cursor = NULL;
		L->index = -1;
	}
}

void moveNext(List L){
	if(L == NULL){
		printf("List error: moveNext is NULL");
		exit(1);
	}
	if(L->cursor != NULL && L-> cursor != L->back){
		L->cursor = L->cursor-> next;
		L->index++;
	}
	else if (L->cursor != NULL && L->cursor == L->back){
		L->cursor = NULL;
		L->index= -1;
	}
}

void prepend(List L, int data){
	if(L == NULL){
		printf("List error: prepend is NULL");
		exit(1);
	}
	Node N = newNode(data);
	if(L-> length == 0){
		L->front = L->back = N;
	} 
	else{
		N->next = L->front;
		L->front->prev = N;
		L->front = N;
	}
	L-> length++;
	L-> index++;
}

void append(List L, int data){
	if(L == NULL){
		printf("List error: append() is NULL");
		exit(1);
	}
	Node N = newNode(data);
	if (L->length == 0){
		L->front = L->back = N;
	} 
	else{
		N->prev = L ->back;
		L->back->next  = N;
		L->back = N;
	}
	L->length++;
}

void insertBefore(List L, int data){
	if(L->length<= 0){
		printf("List error: insertBefore() is empty");
		exit(1);
	}
	if (L->index < 0){
		printf ("List error: insertBefore() when index is less than 0");
		exit(1);
	}
	if (index(L) == 0) {
		prepend(L,data);
	}
	else{
		Node N = newNode(data);
		N->prev = L->cursor->prev;
		N->next = L->cursor;
		L->cursor->prev->next = N;
		L->cursor->prev = N;
		L->index++;
		L->length++;
	}
}

void insertAfter(List L, int data){
	if (L->length <= 0){
		printf("List error: insertAfter() is empty");
		exit(1);
	}
	if (L->index < 0){
		printf ("List error: insertAfter() when index is less than 0");
		exit(1);
	}
	if (index(L) == L->length-1) {
		append(L,data);
	}
	else{
		Node N = newNode(data);
		N->prev = L->cursor;
		N->next = L->cursor->next;
		L->cursor->next->prev = N ;
		L->cursor->next = N;
		L->length++;
	}
}

void deleteFront(List L){
	if(L == NULL){
		printf("List error: deleteFront() is NULL");
		exit(1);
	}
	if(L->length <= 0){
		printf("List error: deleteFront() is empty");
		exit(1);
	}
	Node N = L->front;
	if(L->length == 1){
		L->front = L->back = L->cursor = NULL;
		L->length = 0;
		L->index = -1;
	}
	else{
		 if (L->cursor == L->front){
			L->cursor = NULL;
			L->index = -1;
		}
		else if(L->cursor!= NULL){
			L->index--;
		}
		L->front = L->front->next;
		L->front->prev = NULL;
		L->length--;
	}
	freeNode(&N);
}

void deleteBack(List L){
	if(L == NULL){
		printf("List error: deleteBack() is NULL");
		exit(1);
	}
	if (L->length <= 0){
		printf("List error: deleteBack() is empty");
		exit(1);
	}
	Node N = L->back;
	if(L->length == 1){
		L->front = L->back = L->cursor = NULL;
		L->length = 0;
		L->index = -1;
	}
	else if (L->cursor == L->back){
			L->cursor = NULL;
			L->index = -1;
		}
		else{
		L->back = L->back->prev;
			if (L->back != NULL){
				L->back->next = NULL;
			}
		L->length--;
		}
	freeNode(&N);
}

void delete(List L){
	Node N = NULL;
	if(L == NULL){
		printf("List error: delete() is NULL");
	}
	if(L->length<=0){
		printf("List error: delete() is empty");
		exit(1);
	}
	if(L->index<0){
		printf("List error: delete() when index is less than 0");
		exit(1);
	}
	N = L ->cursor;
	if(L->cursor == L->back){
		deleteBack(L);
		return;
	}
	else if(L->cursor == L->front){
		deleteFront(L);
		return;
	}
	else{
		L->cursor->next->prev = L->cursor->prev;
		L->cursor->prev->next = L->cursor->next;
		L->index=-1;
		L->cursor = NULL;
		L->length--;
		freeNode(&N);
	}
}

void printList(FILE* out, List L){
	if(L->length <= 0){
		return;
	}
	Node N =  NULL;
	if(L == NULL){
		fprintf(out,"List error: printList() is NULL");
		exit(1);
	}
	if(out == NULL){
		printf("List error: printlist() on null file");
		exit(1);
	}
	for(N = L->front ; N!= NULL; N = N->next){
		fprintf(out, "%d", N->data);
	}
}

List copyList(List L){
	List l = newList();
	if (L == NULL){
		printf("List error: copyList() is NULL");
		exit(1);
	}
	if (L->length <= 0){
		return l;
	}
	Node N = L->front;
	while(N!= NULL){
		append(l, N->data);
		N = N->next;
	}
	return l;
}



